src/dm/DispatchManagerActions.cc

```diff
@@ -1878,6 +1906,198 @@ int DispatchManager::detach_nic(int vid, int nic_id, const RequestAttributes& ra
 /* -------------------------------------------------------------------------- */
 /* -------------------------------------------------------------------------- */
 
+int DispatchManager::update_nic(int vid, int nic_id, VirtualMachineTemplate* tmpl,
+        bool append, const RequestAttributes& ra, std::string& error_str)
+{
+    ostringstream oss;
+    oss << "Error updating NIC (" << nic_id << ") for VM " << vid << ": ";
+
+    auto vm = vmpool->get(vid);
+
+    if ( !vm )
+    {
+        oss << "VM does not exist";
+        error_str = oss.str();
+
+        NebulaLog::error("DiM", error_str);
+
+        return -1;
+    }
+
+    VirtualMachineNic *nic = vm->get_nic(nic_id);
+
+    if ( !nic )
+    {
+        oss << "NIC does not exist";
+        error_str = oss.str();
+
+        NebulaLog::error("DiM", error_str);
+
+        return -1;
+    }
+
+    VirtualMachine::VmState  state     = vm->get_state();
+    VirtualMachine::LcmState lcm_state = vm->get_lcm_state();
+
+    bool is_update = false;
+
+    if (state == VirtualMachine::ACTIVE)
+    {
+        if (lcm_state == VirtualMachine::RUNNING)
+        {
+            is_update = true;
+        }
+        else
+        {
+            oss << "Wrong state " << vm->state_str();
+            error_str = oss.str();
+
+            NebulaLog::error("DiM", error_str);
+
+            return -1;
+        }
+    }
+
+    // Get new nic from the template
+    auto nic_va = tmpl->get("NIC");
+
+    if (!nic_va)
+    {
+        oss << "Wrong NIC definition, attribute is missing";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    VectorAttribute* va;
+
+    if (append)
+    {
+        va = nic->vector_attribute()->clone();
+
+        va->merge(nic_va, true);
+    }
+    else
+    {
+        va = nic_va;
+    }
+
+    VirtualMachineNic new_nic(va, nic_id);
+
+    auto rc = vm->nic_update(nic_id, &new_nic, is_update);
+
+    if (is_update && rc > 0)
+    {
+        int net_id = -1;
+
+        vm->set_state(VirtualMachine::HOTPLUG_NIC);
+
+        vm->set_vm_info();
+
+        close_cp_history(vmpool, vm.get(), VMActions::NIC_UPDATE_ACTION, ra);
+
+        nic->vector_value("NETWORK_ID", net_id);
+
+        vmm->updatenic(vm.get(), net_id);
+    }
+
+    if (append)
+    {
+        delete va;
+    }
+
+    vmpool->update(vm.get());
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::recover_nic(int vid, int nic_id, int network_id,
+        string& error_str)
+{
+    ostringstream oss;
+    oss << "Error recovering NIC (" << nic_id << ") for VM " << vid << ": ";
+
+    if (auto vn = vnpool->get(network_id))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.